design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,218 @@
+
+### Non-Goals
+
+This proposal does not attempt to fulfill the following goals:
+
+- Allow the transfer of BIOS configuration between different vendor’s
+  servers or different vendor models
+- Support bulk set of other configurations besides BIOS
+- Provide storage mechanism for BIOS settings
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+When provisioning a large amount of machines with identical settings, our
+customers want to provide a known, validated payload with all the desired
+firmware configuration.
+
+This ensures all the machines will have the identical settings, reducing issues
+caused by unexpected configuration differences.
+
+Since the name/value pair mapping is vendor specific and not intuitive, it
+doesn’t lend itself to creating the input by hand using existing vendor
+documentation, except for potentially a small subset of names. It’s desirable
+to take an existing BIOS config set and modify the current settings to the
+desired settings. The BIOS Registry can be used as a source of documentation
+to modify the settings.
+
+## Design Details
+
+This proposes three new data elements in the BareMetalHost:
+
+- 'currentBIOSSettings' - the current BIOS Settings retrieved from Ironic via
+  the bios API. This dataset includes BIOS Registry information which provides
+  a schema for the settings.
+- 'updateBIOSSettingsTemplate' - a list of writeable BIOS Setting name/value
+  pairs from 'currentBIOSSettings', suitable for setting to a host via Ironic
+- `updateBIOSSettings` - user-defined BIOS Settings to write to host via Ironic.
+  It's assumed that this will be generated from 'updateBIOSSettingsTemplate'
+
+The BIOS Settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the settings
+are updated. Recent changes to Ironic (see the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681)
+added an additional query of the BIOS Registry.
+
+The baremetal-operator can get the BIOS Settings from Ironic after the node
+has been moved to manageable via the `v1/nodes/{node_ident}/bios?detail=True`
+API.  This API provides a list of BIOS Settings (aka `Attributes`) in JSON and each
+setting includes the following data:
+
+- Name
+- Value
+- Registry information
+  - Description
+  - Type (e.g. enum, string, integer, or boolean)
+  - ReadOnly (a boolean indicating whether the attribute is read only)
+  - Immutable (a boolean indicating whether or not the attribute is changeable)
+  - Help text (localized)
+  - Allowable values for enum type attributes
+  - Minimum and maximum values (for numeric attributes)
+  - Minimum and maximum character lengths (for string attributes)
+
+Upon retrieving the BIOS Settigs from Ironic, the baremetal-operator will
+populate the 'currentBIOSSettings` field for each host, represented as a
+ConfigMap with an annotation in BareMetalHostStatus. As there are no fields
```

Comment:
Rewrote this section as we discussed to use a separate CRD for the BIOS settings info read from Ironic.  Depending on the vendor this can be up to 100K.